runtime/lua/vim/lsp/buf.lua

```diff
@@ -96,6 +99,13 @@ local function get_locations(method, opts)
       if #all_items == 1 then
         local item = all_items[1]
         local b = item.bufnr or vim.fn.bufadd(item.filename)
+
+        -- Save position in jumplist
+        vim.cmd("normal! m'")
+        -- Push a new item into tagstack
+        local tagstack = { { tagname = tagname, from = from } }
+        vim.fn.settagstack(vim.fn.win_getid(win), { items = tagstack }, 't')
+
```

Comment:
Note all this logic might be a candidate for vim.jump (https://github.com/neovim/neovim/issues/26782)